content/blog/2023-11-28-context-switching-sprs-on-powerpc.md

```diff
@@ -0,0 +1,443 @@
+Title: Context switching SPRs on PowerPC
+Date: 2023-11-28 08:00:00
+Authors: Benjamin Gray
+Category: Development
+Tags: linux
+
+## Introduction
+
+This article is a dive (well, more of a meander) through some of the PowerPC
+specific aspects of context switching, especially on the Special Purpose
+Register (SPR) handling. It was motivated by my recent work on adding kernel
+support for a hardware feature that interfaces with software through an SPR.
+
+
+## What's a context anyway?
+
+The context we are concerned about in this article is the task context. That's
+all the state that makes up a 'task' in the eyes of the kernel. There are the
+obvious components, like the current register values, thread ID, memory
+mappings, etc., but there are also a few SPRs that get tracked on a per task
+basis. By tracking SPR values for each task context, the kernel can emulate
+per-task support for these SPRs, despite the hardware being per-CPU. The kernel
```

Comment:
I'm not sure I'd talk about per-task SPR save/restore as "emulating per-task support"

Reply:
this aspect of SPRs is no different from GPRs, which also have to be saved and restored. talk about this in terms of the need to save and restore registers rather than "emulate per-task support"

Reply:
or maybe talk about it in terms of emulating per task support but more clearly idk. I get what you're getting at but it's probably strange to an unfamiliar reader